token_dispenser/token_dispenser_lambda.py

```diff
@@ -0,0 +1,176 @@
+
+    :param p12_file_path: Path to the PKCS#12 file (.p12 or .pfx).
+    :param password: Password for the PKCS#12 file (if any).
+    :return: A tuple containing the private key, certificate, and additional certificates.
+    """
+    with open(p12_file_path, "rb") as file:
+        pkcs12_data = file.read()
+
+    # Decode the PKCS#12 file
+    private_key, certificate, additional_certs = load_key_and_certificates(
+        pkcs12_data,
+        password.encode("utf-8") if password else None,
+        backend=default_backend()
+    )
+    return private_key, certificate, additional_certs
+
+def build_cached_cert_file(private_key, certificate):
+    logger = shared_logger()
+    global cached_cert_file
+    try:
+        logger.debug(f"entered build_cached_cert_file")
+        # Serialize private key and certificate into PEM format
+        private_key_pem = private_key.private_bytes(
+            encoding=Encoding.PEM,
+            format=PrivateFormat.PKCS8,
+            encryption_algorithm=NoEncryption()
+        ).decode("utf-8")
+
+        certificate_pem = certificate.public_bytes(Encoding.PEM).decode("utf-8")
+        # Combine private key and certificate into a single file for mutual TLS authentication
+        # The single file is managed by tempfile.NamedTemporaryFile where the library creates an arbitrary file
+        # and be responsible to remove the file while exiting the python virtual machine even if exception happened
+        if cached_cert_file is None or os.path.getsize(cached_cert_file.name) >0:
+            cached_cert_file = NamedTemporaryFile(delete=True)  # Important: delete=True. deletion after python exits
+            cached_cert_file.write(private_key_pem.encode('utf-8'))
+            cached_cert_file.write(certificate_pem.encode('utf-8'))
+            cached_cert_file.flush()
+            logger.info('cached certificate pem and private combination created.')
+        else:
+            logger.info('cached certificate pem and private combination already exists.')
+            cached_cert_file.seek(0)  # reset file pointer
+        return cached_cert_file
+    except Exception as e:
+        logger.error(f"build_cached_cert_file error occurred: {str(e)}")
+        raise e
+
+def get_new_token(client_id:str):
+    """
+    obtain a new token based on client_id and write the newly obtained token into DynamoDB
+    This function try to cache the cert file (a combination of private key and cert str). If the cached cert file exists,
+     use the cached cert to get token.  Otherwise, go through pkc12 download, decode, build cert file and get token
+
+    :param client_id: a required field which is used as key to cache the token in dynamoDB.
+    :return: a json object containing the token data.
+    """
+    global cached_cert_file
+    logger = shared_logger()
+    try:
+        if cached_cert_file is not None and os.path.exists(cached_cert_file.name) and os.path.getsize(cached_cert_file.name) >0:
```

Comment:
Should only need to have `cached_cert_file is not None` since when we create the `NamedTemporaryFile`, we immediately write to it. There shouldn't be a case where the file doesn't exist because we're not multithreading

Reply:
Should only need to have cached_cert_file is not None since when we create the NamedTemporaryFile, we immediately write to it. There shouldn't be a case where the file doesn't exist because we're not multithreading